File: app/src/main/java/ru/ifmo/md/exam1/MainActivity.java
package ru.ifmo.md.exam1;

import android.app.LoaderManager;
import android.content.CursorLoader;
import android.content.Loader;
import android.database.Cursor;
import android.support.v7.app.ActionBarActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.ListView;

import java.util.Random;

import ru.ifmo.md.exam1.provider.song.SongColumns;
import ru.ifmo.md.exam1.provider.song.SongCursor;


public class MainActivity extends ActionBarActivity implements LoaderManager.LoaderCallbacks<Cursor> {
    ArrayAdapter<String> adapter;
    Random rng = new Random(58);

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        adapter = new ArrayAdapter<String>(this, android.R.layout.simple_list_item_1);
        getLoaderManager().restartLoader(0, null, this);
        ((ListView)findViewById(R.id.listView)).setAdapter(adapter);
    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    public void onButtonClick(View view) {

    }

    @Override
    public Loader<Cursor> onCreateLoader(int i, Bundle bundle) {
        return new CursorLoader(this, SongColumns.CONTENT_URI, null, null, null, null);
    }

    @Override
    public void onLoadFinished(Loader<Cursor> cursorLoader, Cursor cursor) {
        if (adapter == null) {
            adapter = new ArrayAdapter<String>(this, android.R.layout.simple_list_item_1);
        }
        adapter.clear();
        while (cursor.moveToNext()) {
            SongCursor itemCursor = new SongCursor(cursor);
            adapter.add(itemCursor.getSong());
        }
        adapter.notifyDataSetChanged();
    }

    @Override
    public void onLoaderReset(Loader<Cursor> cursorLoader) {
        adapter = null;
    }

    public void onButton2Click(View view) {
        getContentResolver().delete(SongColumns.CONTENT_URI, null, null);
    }
}

File: app/src/main/scala/me/volhovm/android/exam2/MainActivity.scala
package me.volhovm.android.exam2

import java.io.{BufferedReader, InputStreamReader}

import android.app.Activity
import android.app.LoaderManager.LoaderCallbacks
import android.content.{AsyncTaskLoader, Intent, Loader}
import android.os.Bundle
import android.util.Log
import android.view.{Menu, MenuItem, View}
import android.widget.{AdapterView, ArrayAdapter, ListView, Toast}
import org.json.JSONArray

class MainActivity extends Activity with LoaderCallbacks[List[PlayList]] {
  private var mPlaylists: List[PlayList] = null
  private var mDatabaseHelper: DatabaseHelper = null
  private var mListView: ListView = null
  private var mListViewAdapter: ArrayAdapter[String] = null

  override def onCreate(savedInstanceState: Bundle): Unit = {
    super.onCreate(savedInstanceState)
    Log.d(this.toString, "In main activity")
    mDatabaseHelper = new DatabaseHelper(this, () => {
      val ins = new BufferedReader(new InputStreamReader(getResources.openRawResource(R.raw.music)))
      val jsonArray: JSONArray = new JSONArray(ins.readLine())
      (for (i <- 0 until jsonArray.length()) yield {
        val currentItem = jsonArray.getJSONObject(i)
        val fullname = currentItem.getString("name")
        val artist = fullname.split("::")(0).trim
        val name = fullname.split("::")(1).trim
        val fullduration = currentItem.getString("duration")
        val duration = Integer.parseInt(fullduration.split(":")(0).trim) * 60 +
          Integer.parseInt(fullduration.split(":")(1).trim)
        new Song(
          name, artist,
          currentItem.getString("url"),
          duration,
          currentItem.getInt("popularity"),
          (for (genre <- 0 until currentItem.getJSONArray("genres").length()) yield {
            currentItem.getJSONArray("genres").getString(genre)
          }).toList,
          currentItem.getInt("year"),
          i
        )
      }).toList
    })
    mListView = new ListView(this)
    mListView.setOnItemClickListener(new AdapterView.OnItemClickListener {
      override def onItemClick(p1: AdapterView[_], p2: View, pos: Int, p4: Long): Unit = {
        val intent = new Intent(MainActivity.this, classOf[PlaylistActivity])
        intent.putExtra("playlist_name", mPlaylists(pos).name)
        startActivity(intent)
      }
    })
    setContentView(mListView)
    getLoaderManager.initLoader(0, null, this).forceLoad()
  }

  override def onCreateOptionsMenu(menu: Menu): Boolean = {
    getMenuInflater.inflate(R.menu.menu_main, menu)
    true
  }

  override def onOptionsItemSelected(item: MenuItem) = item.getItemId match {
    case R.id.action_settings => true
    case R.id.action_refresh =>
      Toast.makeText(this, "TEST", Toast.LENGTH_LONG).show()
      super.onOptionsItemSelected(item)
      true
  }

  override def onCreateLoader(p1: Int, p2: Bundle): Loader[List[PlayList]] =
    new AsyncTaskLoader[List[PlayList]](this) {
      override def loadInBackground(): List[PlayList] = mDatabaseHelper.mWrapper.getAllPlayLists
    }
  override def onLoaderReset(p1: Loader[List[PlayList]]): Unit = ()
  override def onLoadFinished(p1: Loader[List[PlayList]], p2: List[PlayList]): Unit = {
    if (p2.isEmpty) {
      Thread.sleep(2000)
      getLoaderManager.restartLoader(0, null, this)
    }
    mPlaylists = p2
    mListViewAdapter = new ArrayAdapter(this, android.R.layout.simple_list_item_1, mPlaylists.map(_.name).toArray)
    mListView.setAdapter(mListViewAdapter)
    mListViewAdapter.notifyDataSetChanged()
  }
}